examples/assembly/addarrays.s

```diff
@@ -0,0 +1,116 @@
+.globl addarrays_serial
+.type addarrays_serial, %function
+
+# void addarrays_serial(int *res, int *A, int *B, int N)
+# Set res[i] = A[i] + B[i] for i from [0, N)
+# x0 -> res
+# x1 -> A
+# x2 -> B
+# w3 -> N
+# w4 -> loop counter
+
+addarrays_serial:
+    mov w4, wzr
+.loop_begin:
+    cmp w4, w3
+    beq .loop_exit
+    ldr w5, [x1, w4, uxtw 2]
+    ldr w6, [x2, w4, uxtw 2]
+    add w5, w5, w6
+    str w5, [x0, w4, uxtw 2]
+    add w4, w4, 1
+    b .loop_begin
+.loop_exit:
+    ret
+
+.global addarrays_sve
+.type addarrays_sve, %function
+
+# void addarrays_sve (int *restrict res, int *A, int *B, int N)
+# Set res[i] = A[i] + B[i] for i from [0, N) using SVE vectors.
+# x0 -> res
+# x1 -> A
+# x2 -> B
+# x3 -> N
+# x4 -> loop counter
+
+addarrays_sve:
+    mov x4, xzr
+.loop_begin_2:
+    whilelt p0.s, x4, x3
+    b.none .loop_exit_2
+    ld1w z0.s, p0/z, [x1, x4, lsl 2]
+    ld1w z1.s, p0/z, [x2, x4, lsl 2]
+    add z0.s, z0.s, z1.s
+    st1w z0.s, p0, [x0, x4, lsl 2]
+    incw x4
+    b .loop_begin_2
```

Comment:
This needs some basic comments to explain how the scalable stuff works. At least on whilelt and the later incw.

Also the b.none, which on the face of it doesn't make any sense but I see that it branches if no elements of the predicate are set. So that needs a comment too.

Reply:
By which I mean "b.none" almost looks like a branch you'd never take, but clearly that's not the case here.

Reply:
Heh, or maybe the following looks better which doesn't invert the condition ?

```
addvectors_sve:
    mov x4, 0
    b .cond
.loop_body:
    ld1w z0.s, p0/z, [x1, x4, lsl 2]
    ld1w z1.s, p0/z, [x2, x4, lsl 2]
    add z0.s, z0.s, z1.s
    st1w z0.s, p0, [x0, x4, lsl 2]
    incw x4
.cond:
    whilelt p0.s, x4, x3
    b.first .loop_body
    ret
```

Reply:
Yes that looks nicer, I'd still add a comment to explain the decision `b.first` is asking.
```
    // Generate a predicate for the remaining items.
    whilelt p0.s, x4, x3
    // If any elements of that predicate are enabled, run the loop body.
    b.first .loop_body
```

Also:
```
  // Increment x4 by the number of 32 bit words in the vector length.
  incw x4
```

This isn't meant to be an SVE tutorial by any means, so the comments are for the casual reader more than anyone else.